Issue building oneDNN 3.4.4 with CLANG for ARM64 on Windows

# Summary
Some code paths for ARM64 fail to build for Windows when using Clang with MSYS2 ARM64 (CLANGARM64 environment), but no issues with building using Clang with MSYS2 for x86_64 (CLANG64 environment). 

Apparently Aarch64 code paths are building as if the OS is Linux 

# Version
githash: 0cb68d6391b90d12fe5a22f99db21c0624994668

# Environment
* CPU: Snapdragon 8cx Gen 3
* OS version: MINGW64_NT-10.0-22621
* Compiler version: clang version 17.0.6
        Target: aarch64-w64-windows-gnu
        Thread model: posix
        InstalledDir: C:/msys64/clangarm64/bin
* CMake version:  3.28.3

* CMake output log:
```
CMAKE_BUILD_TYPE is unset, defaulting to Release
The C compiler identification is Clang 17.0.6
The CXX compiler identification is Clang 17.0.6
Detecting C compiler ABI info
Detecting C compiler ABI info - done
Check for working C compiler: C:/msys64/clangarm64/bin/cc.exe - skipped
Detecting C compile features
Detecting C compile features - done
Detecting CXX compiler ABI info
Detecting CXX compiler ABI info - done
Check for working CXX compiler: C:/msys64/clangarm64/bin/c++.exe - skipped
Detecting CXX compile features
Detecting CXX compile features - done
DNNL_TARGET_ARCH: AARCH64
DNNL_LIBRARY_NAME: dnnl
Performing Test CMAKE_HAVE_LIBC_PTHREAD
Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
Looking for pthread_create in pthreads
Looking for pthread_create in pthreads - not found
Looking for pthread_create in pthread
Looking for pthread_create in pthread - found
Found Threads: TRUE  
Found OpenMP_C: -fopenmp=libomp (found version "5.1") 
Found OpenMP_CXX: -fopenmp=libomp (found version "5.1") 
Found OpenMP: TRUE (found version "5.1")  
Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
Could NOT find Doxyrest (missing: DOXYREST_EXECUTABLE) 
CMake Warning (dev) at cmake/Sphinx.cmake:25 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  cmake/doc.cmake:19 (include)
  CMakeLists.txt:127 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

Found PythonInterp: C:/msys64/clangarm64/bin/python.exe (found suitable version "3.11.7", minimum required is "2.7") 
Could NOT find Sphinx (missing: SPHINX_EXECUTABLE) 
Found Git: C:/msys64/usr/bin/git.exe (found version "2.43.1") 
Enabled testing coverage: CI
Enabled workload: TRAINING
Enabled primitives: ALL
Enabled primitive CPU ISA: ALL
Enabled primitive GPU ISA: ALL
Enabled GeMM kernels ISA: ALL
Primitive cache is enabled
Graph component is enabled
Graph compiler backend is disabled.
Configuring done (8.7s)
Generating done (0.5s)
```

# Steps to reproduce
Build oneDNN with MSYS2 on a Windows on ARM64 machine in the CLANGARM64 environment.

# Observed behavior
```
-- Build files have been written to: C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/build-CLANGARM64
[15/816] Building CXX object src/common/CMakeFiles/dnnl_common.dir/gemm.cpp.obj
FAILED: src/common/CMakeFiles/dnnl_common.dir/gemm.cpp.obj
C:\msys64\clangarm64\bin\clang++.exe -DDNNL_AARCH64=1 -DDNNL_DLL -DDNNL_DLL_EXPORTS -DDNNL_ENABLE_CPU_ISA_HINTS -DDNNL_ENABLE_ITT_TASKS -DDNNL_ENABLE_MAX_CPU_ISA -DNOMINMAX -DWIN32_LEAN_AND_MEAN -D_WIN -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS -IC:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/build-CLANGARM64/include -IC:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/include -IC:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -fopenmp=libomp -fvisibility-inlines-hidden  -Wall -Wno-unknown-pragmas -fvisibility=internal -O3 -mcpu=generic   -Wno-pass-failed  -O3 -DNDEBUG -std=c++11 -MD -MT src/common/CMakeFiles/dnnl_common.dir/gemm.cpp.obj -MF src\common\CMakeFiles\dnnl_common.dir\gemm.cpp.obj.d -o src/common/CMakeFiles/dnnl_common.dir/gemm.cpp.obj -c C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src/common/gemm.cpp
In file included from C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src/common/gemm.cpp:27:
In file included from C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src/cpu/gemm/gemm.hpp:47:
In file included from C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src/cpu/aarch64/cpu_isa_traits.hpp:35:
C:/Dev/Github/MINGW-packages/mingw-w64-onednn/src/oneDNN-3.3.4/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64.h:52:10: fatal error: 'sys/mman.h' file not found
   52 | #include <sys/mman.h>
      |          ^~~~~~~~~~~~
1 error generated.
[22/816] Building CXX object src/common/CMakeFiles/dnnl_common.dir/inner_product.cpp.obj
ninja: build stopped: subcommand failed.
```

# Expected behavior
oneDNN builds

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64.h
@@ -37,7 +37,7 @@
 #include <unordered_set>
 #include <vector>
 
-#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(__MINGW64__)
+#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(_WIN32)
 #ifndef XBYAK_USE_MMAP_ALLOCATOR
 #define XBYAK_USE_MMAP_ALLOCATOR
 #endif
@@ -46,7 +46,7 @@
 #include <cmath>
 #include <functional>
 
-#if defined(__GNUC__) && !defined(__MINGW64__)
+#if defined(__GNUC__) && !defined(_WIN32)
 #include <cassert>
 #include <stdlib.h>
 #include <sys/mman.h>

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64.h
@@ -37,7 +37,7 @@
 #include <unordered_set>
 #include <vector>
 
-#if defined(__GNUC__) || defined(__APPLE__)
+#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(__MINGW64__)
 #ifndef XBYAK_USE_MMAP_ALLOCATOR
 #define XBYAK_USE_MMAP_ALLOCATOR
 #endif
@@ -46,7 +46,7 @@
 #include <cmath>
 #include <functional>
 
-#if defined(__GNUC__)
+#if defined(__GNUC__) && !defined(__MINGW64__)
 #include <cassert>
 #include <stdlib.h>
 #include <sys/mman.h>

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h
@@ -20,7 +20,7 @@
 static const size_t CSIZE = sizeof(uint32_t);
 
 inline void *AlignedMalloc(size_t size, size_t alignment) {
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__MINGW64__)
   return _aligned_malloc(size, alignment);
 #else
   void *p;
@@ -30,7 +30,7 @@ inline void *AlignedMalloc(size_t size, size_t alignment) {
 }
 
 inline void AlignedFree(void *p) {
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__MINGW64__)
   _aligned_free(p);
 #else
   free(p);
@@ -277,7 +277,7 @@ public:
     default:
       return false;
     }
-#if defined(__GNUC__) || defined(__APPLE__)
+#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(__MINGW64__)
     size_t pageSize = inner::getPageSize();
     size_t iaddr = reinterpret_cast<size_t>(addr);
     size_t roundAddr = iaddr & ~(pageSize - static_cast<size_t>(1));

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_code_array.h
@@ -20,7 +20,7 @@
 static const size_t CSIZE = sizeof(uint32_t);
 
 inline void *AlignedMalloc(size_t size, size_t alignment) {
-#if defined(_MSC_VER) || defined(__MINGW64__)
+#if defined(_WIN32)
   return _aligned_malloc(size, alignment);
 #else
   void *p;
@@ -30,7 +30,7 @@ inline void *AlignedMalloc(size_t size, size_t alignment) {
 }
 
 inline void AlignedFree(void *p) {
-#if defined(_MSC_VER) || defined(__MINGW64__)
+#if defined(_WIN32)
   _aligned_free(p);
 #else
   free(p);
@@ -277,7 +277,7 @@ public:
     default:
       return false;
     }
-#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(__MINGW64__)
+#if (defined(__GNUC__) || defined(__APPLE__)) && !defined(_WIN32)
     size_t pageSize = inner::getPageSize();
     size_t iaddr = reinterpret_cast<size_t>(addr);
     size_t roundAddr = iaddr & ~(pageSize - static_cast<size_t>(1));

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
@@ -20,7 +20,7 @@ enum { DEFAULT_MAX_CODE_SIZE = 4096 };
 namespace inner {
 
 inline size_t getPageSize() {
-#if defined(__GNUC__) && !defined(__MINGW64__)
+#if defined(__GNUC__) && !defined(_WIN32)
   static const size_t pageSize = sysconf(_SC_PAGESIZE);
 #else
   static const size_t pageSize = 4096;

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
@@ -20,7 +20,7 @@ enum { DEFAULT_MAX_CODE_SIZE = 4096 };
 namespace inner {
 
 inline size_t getPageSize() {
-#if defined (__GNUC__) && !defined(__MINGW64__)
+#if defined(__GNUC__) && !defined(__MINGW64__)
   static const size_t pageSize = sysconf(_SC_PAGESIZE);
 #else
   static const size_t pageSize = 4096;

--- a/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
+++ b/src/cpu/aarch64/xbyak_aarch64/xbyak_aarch64/xbyak_aarch64_inner.h
@@ -20,7 +20,7 @@ enum { DEFAULT_MAX_CODE_SIZE = 4096 };
 namespace inner {
 
 inline size_t getPageSize() {
-#ifdef __GNUC__
+#if defined (__GNUC__) && !defined(__MINGW64__)
   static const size_t pageSize = sysconf(_SC_PAGESIZE);
 #else
   static const size_t pageSize = 4096;

--- a/tests/benchdnn/common.cpp
+++ b/tests/benchdnn/common.cpp
@@ -328,7 +328,7 @@ bool skip_start(res_t *res, int idx) {
     return false;
 }
 
-#if defined(_WIN32) && !defined(__GNUC__)
+#if defined(_WIN32)
 #include <windows.h>
 #define PATH_MAX MAX_PATH
 static char *dirname(char *path) {

--- a/tests/benchdnn/dnnl_common.cpp
+++ b/tests/benchdnn/dnnl_common.cpp
@@ -839,7 +839,7 @@ bool is_f64_supported(const dnnl_engine_t &engine) {
     return false;
 }
 
-#if defined(_WIN32) && !defined(__GNUC__)
+#if defined(_WIN32)
 #include "windows.h"
 
 static size_t get_cpu_ram_size() {

